src/main/kotlin/Main.kt

```diff
@@ -96,6 +106,99 @@ private fun executeTests(tests: Collection<Test>): Iterable<ResultContext> {
     return fullResults
 }
 
+private fun generateTests(): Collection<Test> {
+    val allRelations = ProofSearcher().addAllProofs().findNewRelations(RelationLoader.relations)
+    //ProofLimiter(3).limit(allRelations)
+    return TestGenerator().addAllTests().generateTests(allRelations)
+}
+
+
+val OPERATORS = listOf("||", "\\\\", "&&", "consistency:", "refinement:")
+private fun sortTests(engine: EngineConfiguration, tests: Collection<Test>) : Collection<Test> {
+    var out = ArrayList(tests)
+
+    if (engine.queryComplexity != null) { //Query Complexity
+        val upper: Int; val lower: Int
+        if (engine.queryComplexity.size >= 2) {
+            upper = engine.queryComplexity[1]
+            lower = engine.queryComplexity.first()
+        } else {
+            upper = engine.queryComplexity.firstOrNull() ?: Int.MAX_VALUE
+            lower = 0
+        }
+
+        if (upper < lower)
+            throw Exception("The upper bound for `queryComplexity` can't be less than the lower bound")
+
+        out = ArrayList(out.filter { x ->
+            x.queries().all { y ->
+                var c = 0
+                for (op in OPERATORS) {
+                    c += y.split(op)
+                        .dropLastWhile { it.isEmpty() }
+                        .toTypedArray().size - 1
+                }
+                c in lower..upper
+            }
+        })
+    }
+
+    if (engine.testCount != null) {  //Count
+        out = when (engine.testSorting) {
+            Sorting.FILO -> ArrayList(out.takeLast(engine.testCount))
+            Sorting.FIFO -> ArrayList(out.take(engine.testCount))
+            Sorting.RoundRobin -> getEqualTests(out, engine.testCount)
+            Sorting.Random, null -> ArrayList(out.shuffled().take(engine.testCount))
+        }
+    }
+    return out
+}
+
+private fun getEqualTests(tests: Collection<Test>, count: Int): ArrayList<Test> {
+    val map: HashMap<Pair<String, String>, ArrayList<Test>> = HashMap()
+    tests.forEach { x -> map.getOrPut(Pair(x.type, x.testSuite)) { ArrayList() }.add(x) }
+    println(map.keys)
+    val out = ArrayList<Test>()
+    for (test in map.values.toList()) {
+        out.addAll(test.take(count / map.keys.size))
+    }
+    return out
+}
+
+private fun saveTests(engineName: String, path: String, tests: Collection<Test>) {
+    val file = File(Paths.get(path).absolutePathString()).normalize()
+    println("Saving the tests for $engineName in $file")
+    val writer = file.writer(Charset.defaultCharset())
+    val initStr = "Tests generated for engine '$engineName' on ${SimpleDateFormat("dd/M/yyyy hh:mm:ss").format(Date())}"
+    writer.write(initStr + "\n")
```

Comment:
Some platforms writes newlines different, e.g. "\n" and "\r\n" (With carriage return for windows). Instead we should use the system wide ``System.lineSeparator()`` property.